com.unity.multiplayer.mlapi/Runtime/Timing/NetworkTime.cs

```diff
@@ -0,0 +1,155 @@
+        /// </summary>
+        public int Tick => m_Tick;
+
+        /// <summary>
+        /// Gets the tickrate of the system of this <see cref="NetworkTime"/>.
+        /// </summary>
+        public int TickRate => m_TickRate;
+
+        /// <summary>
+        /// Creates a new instance of the <see cref="NetworkTime"/> struct.
+        /// </summary>
+        /// <param name="tickRate">The tickrate.</param>
+        public NetworkTime(int tickRate)
+        {
+            m_TickRate = tickRate;
+            m_TickInterval = 1f / m_TickRate; // potential floating point precision issue, could result in different interval on different machines
+            m_TickDurationOffset = 0;
+            m_Tick = 0;
+        }
+
+        /// <summary>
+        /// Creates a new instance of the <see cref="NetworkTime"/> struct.
+        /// </summary>
+        /// <param name="tickRate">The tickrate.</param>
+        /// <param name="tick">The time will be created with a value where this many tick have already passed.</param>
+        /// <param name="tickDurationOffset">Can be used to create a <see cref="NetworkTime"/> with a non fixed time value by adding an offset to the given tick value.</param>
+        public NetworkTime(int tickRate, int tick, float tickDurationOffset = 0f)
+            : this(tickRate)
+        {
+            Assert.IsTrue(tickDurationOffset < 1f / tickRate);
+
+            m_Tick = tick;
+            m_TickDurationOffset = tickDurationOffset;
+        }
+
+        /// <summary>
+        /// Creates a new instance of the <see cref="NetworkTime"/> struct.
+        /// </summary>
+        /// <param name="tickRate">The tickrate.</param>
+        /// <param name="time">The time value as a float.</param>
+        public NetworkTime(int tickRate, float time)
+            : this(tickRate)
+        {
+            this += time;
+
+            // This is due to floating point precision issues.
+            if (m_TickDurationOffset < 0)
+            {
+                this += m_TickDurationOffset;
+            }
+        }
+
+
+        /// <summary>
+        /// Converts the network time into a fixed time value.
+        /// </summary>
+        /// <returns>A <see cref="NetworkTime"/> where Time is the FixedTime value of this instance.</returns>
+        public NetworkTime ToFixedTime()
+        {
+            return new NetworkTime(m_TickRate, m_Tick);
+        }
+
+        public static NetworkTime operator -(NetworkTime a, NetworkTime b)
+        {
+            Assert.AreEqual(a.TickRate, b.TickRate, $"NetworkTimes must have same TickRate to subtract.");
+
+            int tick = a.Tick - b.Tick;
+            float tickDuration = a.m_TickDurationOffset - b.m_TickDurationOffset;
+
+            if (tickDuration < 0)
+            {
+                tick--;
+                tickDuration += a.m_TickInterval;
+            }
+
+            return new NetworkTime(a.TickRate, tick, tickDuration);
```

Comment:
Are you at all worried about the having to instantiate / cleanup lots of NetworkTime's?

Reply:
There is absolutely nothing we can do here to prevent this. NetworkTime is a value type, we have to return a new value. Yes this is a stack alloc. But `int i += 1;` also creates a stack alloc.